File: moonpoison/binary_tree-dfs/1448.cpp
#include<bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
public:
    int ans=1;
    int goodNodes(TreeNode* root)
    {
        dfs(root, root->val);
        return ans;
    }
    void dfs(TreeNode* root, int num)
    {
        if(root->left!=nullptr)
        {
            if(root->left->val>=num) ans++;
            dfs(root->left, max(num, root->left->val));   
        }
        if(root->right!=nullptr)
        {
            if(root->right->val>=num) ans++;
            dfs(root->right, max(num, root->right->val));
        }
    }
};

File: ice0208/binary-tree-dfs/1448.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def go_down(self, cursor: TreeNode, max_value: int):
        if cursor.val >= max_value:
            self.cnt += 1
        max_value = max(max_value, cursor.val)

        if cursor.left:
            self.go_down(cursor.left, max_value)
        if cursor.right:
            self.go_down(cursor.right, max_value)

    def goodNodes(self, root: TreeNode) -> int:
        self.cnt = 0

        if root:
            self.go_down(root, -(10**4))

        return self.cnt
